File: FormalATM/src/gov/nasa/larcfm/ACCoRD/AlertThresholds.java
/*
 * Copyright (c) 2015-2016 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */
package gov.nasa.larcfm.ACCoRD;

import java.util.List;

import gov.nasa.larcfm.Util.Interval;
import gov.nasa.larcfm.Util.Units;
import gov.nasa.larcfm.Util.Vect3;
import gov.nasa.larcfm.Util.Velocity;
import gov.nasa.larcfm.Util.f;

public class AlertThresholds {

  private Detection3D detector_; // State-based detector  
  private double alerting_time_; // Alerting_time
  // If alerting_time > 0, alert is based on detection
  // If alerting_time = 0, alert is based on violation
  private Interval spread_trk_; // Alert when track band within spread interval. 
  private Interval spread_gs_;  // Alert when ground speed band within spread interval.
  private Interval spread_vs_;  // Alert when vertical speed band within speed interval.
  private Interval spread_alt_; // Alert when altitude  band within spread interval.
  // Spread intervals must include 0 and the low and up values are relative to ownship's state. 


  /** 
   * Creates an alert threholds object. Parameter det is an arbitrary detector,
   * tthr is a non-negative alerting time (possibly positive infinity). 
   */
  public AlertThresholds(Detection3D det, double tthr) {
    detector_ = det.copy();
    alerting_time_ = Math.max(0,tthr);
    spread_trk_ = new Interval(0,0);
    spread_gs_ = new Interval(0,0);
    spread_vs_ = new Interval(0,0);
    spread_alt_ = new Interval(0,0);
  }

  public AlertThresholds(AlertThresholds athr) {
    detector_ = athr.detector_.copy();
    alerting_time_ = athr.alerting_time_;
    spread_trk_ = athr.spread_trk_;
    spread_gs_ = athr.spread_gs_;
    spread_vs_ = athr.spread_vs_;
    spread_alt_ = athr.spread_alt_;    
  }

  private AlertThresholds() {
    detector_ = null;
    alerting_time_ = Double.NaN;
    spread_trk_ = Interval.EMPTY;
    spread_gs_ = Interval.EMPTY;
    spread_vs_ = Interval.EMPTY;
    spread_alt_ = Interval.EMPTY;
  }

  public static final AlertThresholds INVALID = new AlertThresholds();

  public boolean isValid() {
    return detector_ != null && !Double.isNaN(alerting_time_) &&
        !spread_trk_.isEmpty() && !spread_gs_.isEmpty() &&
        !spread_vs_.isEmpty() && !spread_alt_.isEmpty();
  }

  /**
   * Return detector.
   */
  public Detection3D getDetector() {
    return detector_;
  }

  /**
   * Set detector.
   */
  public void setDetector(Detection3D det) {
    this.detector_ = det.copy();
  }

  /**
   * Return alerting time in seconds.
   */
  public double getAlertingTime() {
    return alerting_time_;
  }

  /**
   * Set alerting time. Parameter val is a non-negative value, possibly positive infinity, given in seconds.
   */
  public void setAlertingTime(double val) {
    alerting_time_ = Math.abs(val);
  }

  /**
   * Get left track, in internal units [rad], relative to ownship's track, for alerting.
   */
  public double getSpreadLeftTrack() {
    return spread_trk_.low;
  }

  /**
   * Get left track, in given units [u], relative to ownship's track, for alerting.
   */
  public double getSpreadLeftTrack(String u) {
    return Units.to(u,getSpreadLeftTrack());
  }  

  /**
   * Get right track, in internal units [rad], relative to ownship's track, for alerting.
   */
  public double getSpreadRightTrack() {
    return spread_trk_.up;
  }

  /**
   * Get right track, in given units [u], relative to ownship's track, for alerting.
   */
  public double getSpreadRightTrack(String u) {
    return Units.to(u,getSpreadRightTrack());
  }  

  /** 
   * Set left/right tracks, relative to ownship's track, for alerting. 
   * Tracks are specified in internal units [rad]. Left value is expected to be in [-pi,0].
   * Right value is expected to be in [0,pi] 
   */
  public void setSpreadTrack(double left, double right) {
    left = -Math.abs(left);
    right = Math.abs(right);
    if (left >= -Math.PI && right <= Math.PI) {
      spread_trk_ = new Interval(left,right);
    }     
  }

  /** 
   * Set left/right tracks, relative to ownship's track, for alerting. 
   * Tracks are specified in given units [u]. Left value is expected to be in [-pi,0] [u].
   * Right value is expected to be in [0,pi] [u]. 
   */
  public void setSpreadTrack(double left, double right, String u) {
    setSpreadTrack(Units.from(u,left),Units.from(u,right));
  }

  /**
   * Get below ground speed, in internal units [m/s], relative to ownship's ground speed, for alerting.
   */
  public double getSpreadBelowGroundSpeed() {
    return spread_gs_.low;
  }

  /**
   * Get below ground speed, in given units [u], relative to ownship's ground speed, for alerting.
   */
  public double getSpreadBelowGroundSpeed(String u) {
    return Units.to(u,getSpreadBelowGroundSpeed());
  }  

  /**
   * Get above ground speed, in internal units [m/s], relative to ownship's ground speed, for alerting.
   */
  public double getSpreadAboveGroundSpeed() {
    return spread_gs_.up;
  }

  /**
   * Get above ground speed, in given units [u], relative to ownship's ground speed, for alerting.
   */
  public double getSpreadAboveGroundSpeed(String u) {
    return Units.to(u,getSpreadAboveGroundSpeed());
  }  

  /** 
   * Set below/above ground speeds, relative to ownship's ground speed, for alerting. 
   * Ground speeds are specified in internal units [m/s]. Below value is expected to be non-positive.
   * Above value is expected to be non-negative.
   */
  public void setSpreadGroundSpeed(double below, double above) {
    below = -Math.abs(below);
    above = Math.abs(above);
    spread_gs_ = new Interval(below,above);
  }

  /** 
   * Set below/above ground speeds, relative to ownship's ground speed, for alerting. 
   * Ground speeds are specified in given units [u]. Below value is expected to be non-positive.
   * Above value is expected to be non-negative.
   */
  public void setSpreadGroundSpeed(double below, double above, String u) {
    setSpreadGroundSpeed(Units.from(u,below),Units.from(u,above));
  }

  /**
   * Get below vertical speed, in internal units [m/s], relative to ownship's vertical speed, for alerting.
   */
  public double getSpreadBelowVerticalSpeed() {
    return spread_vs_.low;
  }

  /**
   * Get below vertical speed, in given units [u], relative to ownship's vertical speed, for alerting.
   */
  public double getSpreadBelowVerticalSpeed(String u) {
    return Units.to(u,getSpreadBelowVerticalSpeed());
  }  

  /**
   * Get above vertical speed, in internal units [m/s], relative to ownship's vertical speed, for alerting.
   */
  public double getSpreadAboveVerticalSpeed() {
    return spread_vs_.up;
  }

  /**
   * Get above vertical speed, in given units [u], relative to ownship's vertical speed, for alerting.
   */
  public double getSpreadAboveVerticalSpeed(String u) {
    return Units.to(u,getSpreadAboveVerticalSpeed());
  }  

  /** 
   * Set below/above vertical speeds, relative to ownship's vertical speed, for alerting. 
   * Vertical speeds are specified in internal units [m/s]. Below value is expected to be non-positive.
   * Above value is expected to be non-negative.
   */
  public void setSpreadVerticaSpeed(double below, double above) {
    below = -Math.abs(below);
    above = Math.abs(above);
    spread_vs_ = new Interval(below,above);
  }

  /** 
   * Set below/above vertical speeds, relative to ownship's vertical speed, for alerting. 
   * Vertical speeds are specified in given units [u]. Below value is expected to be non-positive.
   * Above value is expected to be non-negative.
   */
  public void setSpreadVerticalSpeed(double below, double above, String u) {
    setSpreadVerticaSpeed(Units.from(u,below),Units.from(u,above));
  }

  /**
   * Get below altitude, in internal units [m], relative to ownship's altitude, for alerting.
   */
  public double getSpreadBelowAltitude() {
    return spread_alt_.low;
  }

  /**
   * Get below altitude, in given units [u], relative to ownship's altitude, for alerting.
   */
  public double getSpreadBelowAltitude(String u) {
    return Units.to(u,getSpreadBelowAltitude());
  }  

  /**
   * Get above altitude, in internal units [m], relative to ownship's altitude, for alerting.
   */
  public double getSpreadAboveAltitude() {
    return spread_alt_.up;
  }

  /**
   * Get above altitude, in given units [u], relative to ownship's altitude, for alerting.
   */
  public double getSpreadAboveAltitude(String u) {
    return Units.to(u,getSpreadAboveAltitude());
  }  

  /** 
   * Set below/above altitudes, relative to ownship's altitude, for alerting. 
   * Altitudes are specified in internal units [m]. Below value is expected to be non-positive.
   * Above value is expected to be non-negative.
   */
  public void setSpreadAltitude(double below, double above) {
    below = -Math.abs(below);
    above = Math.abs(above);
    spread_alt_ = new Interval(below,above);
  }

  /** 
   * Set below/above altitudes, relative to ownship's altitude, for alerting. 
   * Altitudes are specified in given units [u]. Below value is expected to be non-positive.
   * Above value is expected to be non-negative.
   */
  public void setSpreadAltitude(double below, double above, String u) {
    setSpreadAltitude(Units.from(u,below),Units.from(u,above));
  }

  /** 
   * Return true if and only if threshold values, defining an alerting level, are violated.
   */ 
  public boolean alerting(DaidalusParameters parameters, TrafficState own, TrafficState ac, 
      TrafficState mu_ac, int turning, int accelerating, int climbing) {
    if (isValid()) {
      Vect3 so = own.get_s();
      Velocity vo = own.get_v();
      Vect3 si = ac.get_s();
      Velocity vi = ac.get_v();
      if (detector_.violation(so,vo,si,vi)) {
        return true;
      }
      ConflictData det = detector_.conflictDetection(so,vo,si,vi,0,alerting_time_);
      if (det.conflict()) {
        return true;
      }
      if (!spread_trk_.isSingle() || !spread_gs_.isSingle() ||
          !spread_vs_.isSingle() || !spread_alt_.isSingle()) {
        KinematicBands bands = new KinematicBands(detector_); 
        bands.setOwnship(own);
        bands.setParameters(parameters);
        bands.setMostUrgentAircraft(mu_ac);
        bands.disableRecoveryBands();
        bands.setAlertingTime(0);
        bands.setLookaheadTime(alerting_time_);
        if (!spread_trk_.isSingle()) {
          bands.setLeftTrack(turning <= 0 ? spread_trk_.low : 0);
          bands.setRightTrack(turning >= 0 ? spread_trk_.up : 0);       
          if (bands.trk_band().kinematic_conflict(bands.core(),ac)) {
            return true;
          }
        }
        if (!spread_gs_.isSingle()) {
          bands.setBelowAboveGroundSpeed(accelerating <= 0 ? spread_gs_.low : 0, 
              accelerating >= 0 ? spread_gs_.up : 0);
          if (bands.gs_band().kinematic_conflict(bands.core(),ac)) {
            return true;
          }
        }
        if (!spread_vs_.isSingle()) {
          bands.setBelowAboveVerticalSpeed(climbing <= 0 ? spread_vs_.low : 0, 
              climbing >= 0 ? spread_vs_.up : 0);
          if (bands.vs_band().kinematic_conflict(bands.core(),ac)) {
            return true;
          }
        }
        if (!spread_alt_.isSingle()) {
          bands.setBelowAboveAltitude(climbing <= 0 ? spread_alt_.low : 0, 
              climbing >= 0 ? spread_alt_.up : 0);
          if (bands.alt_band().kinematic_conflict(bands.core(),ac)) {
            return true;
          }
        }
      }
    }
    return false;
  }

  public String toString() {
    return detector_.toString()+", Alerting Time: "+f.Fm1(alerting_time_)+
        " [s], Spread Track: "+spread_trk_.toStringUnits("deg")+
        ", Spread Ground Speed: "+spread_gs_.toStringUnits("knot")+
        ", Spread Vertical Speed: "+spread_vs_.toStringUnits("fpm")+
        ", Spread Altitude: "+spread_alt_.toStringUnits("ft");
  }

  public static String listToString(List<AlertThresholds> l) {
    String s = "";
    for (int i=0; i < l.size(); ++i) {
      s += "Level "+(i+1)+": "+l.get(i).toString()+"\n";
    }
    return s;
  }

  public String toPVS(int prec) {
    return "(# wcv:= "+detector_.toPVS(prec)+
        ", lookahead:= "+f.Fm1(alerting_time_)+
        ", spread_trk:= "+spread_trk_.toPVS(prec)+
        ", spread_gs:= "+spread_gs_.toPVS(prec)+
        ", spread_vs:= "+spread_vs_.toPVS(prec)+
        ", spread_alt:= "+spread_alt_.toPVS(prec)+
        " #)"; 
  }

  public static String listToPVS(List<AlertThresholds> l, int prec) {
    String s = "(: ";
    boolean first = true;
    for (int i=0; i < l.size(); ++i) {
      if (first) {
        first = false;
      } else {
        s += ",";
      }
      s += l.get(i).toPVS(prec);
    }
    return s+" :)";
  }
}

File: C++/src/ACCoRD/AlertThresholds.cpp
/*
 * Copyright (c) 2015-2017 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */
#include "AlertThresholds.h"
#include "Vect3.h"
#include "Util.h"
#include "Velocity.h"
#include "Detection3D.h"
#include "ConflictData.h"
#include "KinematicMultiBands.h"
#include "format.h"
#include "KinematicBandsParameters.h" // REMOVE ME!  Circular reference!

namespace larcfm {

/**
 * Creates an alert threholds object. Parameter det is a detector,
 * alerting_time is a non-negative alerting time (possibly positive infinity),
 * early_alerting_time is a early alerting time >= at (for maneuver guidance),
 * region is the type of guidance
 */
AlertThresholds::AlertThresholds(const Detection3D* detector,
    double alerting_time, double early_alerting_time,
    BandsRegion::Region region) {
  detector_ = detector == NULL ? NULL :detector->copy();
  alerting_time_ = std::abs(alerting_time);
  early_alerting_time_ = Util::max(alerting_time_,early_alerting_time);
  region_ = region;
  spread_trk_ = 0;
  spread_gs_ = 0;
  spread_vs_ = 0;
  spread_alt_ = 0;
  units_["alerting_time"] = "s";
  units_["early_alerting_time"] = "s";
  units_["spread_trk"] = "deg";
  units_["spread_gs"] = "knot";
  units_["spread_vs"] = "fpm";
  units_["spread_alt"] = "ft";
}

AlertThresholds::AlertThresholds(const AlertThresholds& athr) {
  detector_ = athr.isValid() ? athr.detector_->copy() : NULL;
  alerting_time_ = athr.alerting_time_;
  early_alerting_time_ = athr.early_alerting_time_;
  region_ = athr.region_;
  spread_trk_ = athr.spread_trk_;
  spread_gs_ = athr.spread_gs_;
  spread_vs_ = athr.spread_vs_;
  spread_alt_ = athr.spread_alt_;
  units_ = athr.units_;
}

AlertThresholds::AlertThresholds() {
  detector_ = NULL;
  alerting_time_ = 0;
  early_alerting_time_ = 0;
  region_ = BandsRegion::UNKNOWN;
  spread_trk_ = 0;
  spread_gs_ = 0;
  spread_vs_ = 0;
  spread_alt_ = 0;
  units_["alerting_time"] = "s";
  units_["early_alerting_time"] = "s";
  units_["spread_trk"] = "deg";
  units_["spread_gs"] = "m/s";
  units_["spread_vs"] = "m/s";
  units_["spread_alt"] = "m";
}

const AlertThresholds AlertThresholds::INVALID = AlertThresholds();

bool AlertThresholds::isValid() const {
  return detector_ != NULL && region_ != BandsRegion::UNKNOWN;
}

AlertThresholds::~AlertThresholds() {
  delete detector_;
}

AlertThresholds& AlertThresholds::operator=(const AlertThresholds& athr) {
  delete detector_;
  detector_ = athr.isValid() ? athr.detector_->copy() : NULL;
  alerting_time_ = athr.alerting_time_;
  early_alerting_time_ = athr.early_alerting_time_;
  region_ = athr.region_;
  spread_trk_ = athr.spread_trk_;
  spread_gs_ = athr.spread_gs_;
  spread_vs_ = athr.spread_vs_;
  spread_alt_ = athr.spread_alt_;
  units_ = athr.units_;
  return *this;
}

/**
 * Return detector.
 */
Detection3D* AlertThresholds::getDetectorRef() const {
  return detector_;
}

/**
 * Set detector.
 */
void AlertThresholds::setDetector(const Detection3D* det) {
  delete detector_;
  detector_ = det != NULL ? det->copy() : NULL;
}

/**
 * Return alerting time in seconds.
 */
double AlertThresholds::getAlertingTime() const {
  return alerting_time_;
}

/**
 * Return alerting time in specified units.
 */
double AlertThresholds::getAlertingTime(const std::string& u) const {
  return Units::to(u,alerting_time_);
}

/**
 * Set alerting time in seconds. Alerting time is non-negative number.
 */
void AlertThresholds::setAlertingTime(double val) {
  alerting_time_ = std::abs(val);
}

/**
 * Set alerting time in specified units. Alerting time is non-negative number.
 */
void AlertThresholds::setAlertingTime(double t, const std::string& u) {
  setAlertingTime(Units::from(u,t));
  units_["alerting_time"] = u;
}

/**
 * Return early alerting time in seconds.
 */
double AlertThresholds::getEarlyAlertingTime() const {
  return early_alerting_time_;
}

/**
 * Return early alerting time in specified units.
 */
double AlertThresholds::getEarlyAlertingTime(const std::string& u) const {
  return Units::to(u,early_alerting_time_);
}

/**
 * Set early alerting time in seconds. Early alerting time is a positive number >= alerting time
 */
void AlertThresholds::setEarlyAlertingTime(double t) {
  early_alerting_time_ = std::abs(t);
}

/**
 * Set early alerting time in specified units. Early alerting time is a positive number >= alerting time
 */
void AlertThresholds::setEarlyAlertingTime(double t, const std::string& u) {
  setEarlyAlertingTime(Units::from(u,t));
  units_["early_alerting_time"] = u;
}

/**
 * Return guidance region.
 */
BandsRegion::Region AlertThresholds::getRegion() const {
  return region_;
}

/**
 * Set guidance region.
 */
void AlertThresholds::setRegion(BandsRegion::Region region) {
  region_ = region;
}

/**
 * Get track spread in internal units [rad]. Spread is relative to ownship's track
 */
double AlertThresholds::getTrackSpread() const {
  return spread_trk_;
}

/**
 * Get track spread in given units [u]. Spread is relative to ownship's track
 */
double AlertThresholds::getTrackSpread(const std::string& u) const {
  return Units::to(u,spread_trk_);
}

/**
 * Set track spread in internal units. Spread is relative to ownship's track and is expected
 * to be in [0,pi].
 */
void AlertThresholds::setTrackSpread(double spread) {
  spread_trk_ = std::abs(Util::to_pi(spread));
}

/**
 * Set track spread in given units. Spread is relative to ownship's track and is expected
 * to be in [0,pi] [u].
 */
void AlertThresholds::setTrackSpread(double spread, const std::string& u) {
  setTrackSpread(Units::from(u,spread));
  units_["spread_trk"] = u;
}

/**
 * Get ground speed spread in internal units [m/s]. Spread is relative to ownship's ground speed
 */
double AlertThresholds::getGroundSpeedSpread() const {
  return spread_gs_;
}

/**
 * Get ground speed spread in given units. Spread is relative to ownship's ground speed
 */
double AlertThresholds::getGroundSpeedSpread(const std::string& u) const {
  return Units::to(u,spread_gs_);
}

/**
 * Set ground speed spread in internal units [m/s]. Spread is relative to ownship's ground speed and is expected
 * to be non-negative
 */
void AlertThresholds::setGroundSpeedSpread(double spread) {
  spread_gs_ = std::abs(spread);
}

/**
 * Set ground speed spread in given units. Spread is relative to ownship's ground speed and is expected
 * to be non-negative
 */
void AlertThresholds::setGroundSpeedSpread(double spread, const std::string& u) {
  setGroundSpeedSpread(Units::from(u,spread));
  units_["spread_gs"] = u;
}

/**
 * Get vertical speed spread in internal units [m/s]. Spread is relative to ownship's vertical speed
 */
double AlertThresholds::getVerticalSpeedSpread() const {
  return spread_vs_;
}

/**
 * Get vertical speed spread in given units. Spread is relative to ownship's vertical speed
 */
double AlertThresholds::getVerticalSpeedSpread(const std::string& u) const {
  return Units::to(u,spread_vs_);
}

/**
 * Set vertical speed spread in internal units [m/s]. Spread is relative to ownship's vertical speed and is expected
 * to be non-negative
 */
void AlertThresholds::setVerticalSpeedSpread(double spread) {
  spread_vs_ = std::abs(spread);
}

/**
 * Set vertical speed spread in given units. Spread is relative to ownship's vertical speed and is expected
 * to be non-negative
 */
void AlertThresholds::setVerticalSpeedSpread(double spread, const std::string& u) {
  setVerticalSpeedSpread(Units::from(u,spread));
  units_["spread_vs"] = u;
}

/**
 * Get altitude spread in internal units [m]. Spread is relative to ownship's altitude
 */
double AlertThresholds::getAltitudeSpread() const {
  return spread_alt_;
}

/**
 * Get altitude spread in given units. Spread is relative to ownship's altitude
 */
double AlertThresholds::getAltitudeSpread(const std::string& u) const {
  return Units::to(u,spread_alt_);
}

/**
 * Set altitude spread in internal units [m]. Spread is relative to ownship's altitude and is expected
 * to be non-negative
 */
void AlertThresholds::setAltitudeSpread(double spread) {
  spread_alt_ = std::abs(spread);
}

/**
 * Set altitude spread in given units. Spread is relative to ownship's altitude and is expected
 * to be non-negative
 */
void AlertThresholds::setAltitudeSpread(double spread, const std::string& u) {
  setAltitudeSpread(Units::from(u,spread));
  units_["spread_alt"] = u;
}

std::string AlertThresholds::toString() const {
  return  (detector_ == NULL ? "INVALID_DETECTOR" : detector_->toString())+
      ", alerting_time = "+Units::str(getUnits("alerting_time"),alerting_time_)+
      ", early_alerting_time = "+Units::str(getUnits("early_alerting_time"),early_alerting_time_)+
      ", region = "+BandsRegion::to_string(region_)+
      ", spread_trk = "+Units::str(getUnits("spread_trk"),spread_trk_)+
      ", spread_gs = "+Units::str(getUnits("spread_gs"),spread_gs_)+
      ", spread_vs = "+Units::str(getUnits("spread_vs"),spread_vs_)+
      ", spread_alt = "+Units::str(getUnits("spread_alt"),spread_alt_);
}

std::string AlertThresholds:: toPVS(int prec) const {
  return "(# wcv:= "+(detector_ == NULL ? "INVALID_DETECTOR" :detector_->toPVS(prec))+
      ", alerting_time:= "+Fm1(alerting_time_)+
      ", early_alerting_time:= "+Fm1(early_alerting_time_)+
      ", region:= "+BandsRegion::to_string(region_)+
      ", spread_trk:= ("+FmPrecision(spread_trk_,prec)+","+FmPrecision(spread_trk_,prec)+")"+
      ", spread_gs:= ("+FmPrecision(spread_gs_,prec)+","+FmPrecision(spread_gs_,prec)+")"+
      ", spread_vs:= ("+FmPrecision(spread_vs_,prec)+","+FmPrecision(spread_vs_,prec)+")"+
      ", spread_alt:= ("+FmPrecision(spread_alt_,prec)+","+FmPrecision(spread_alt_,prec)+")"+
      " #)";
}

ParameterData AlertThresholds::getParameters() const {
  ParameterData p;
  updateParameterData(p);
  return p;
}

void AlertThresholds::updateParameterData(ParameterData& p) const {
  p.set("region",BandsRegion::to_string(region_));
  if (detector_ != NULL) {
    p.set("detector",detector_->getIdentifier());
  }
  p.setInternal("alerting_time",alerting_time_,getUnits("alerting_time"));
  p.setInternal("early_alerting_time",early_alerting_time_,getUnits("early_alerting_time"));
  p.setInternal("spread_trk",spread_trk_,getUnits("spread_trk"));
  p.setInternal("spread_gs",spread_gs_,getUnits("spread_gs"));
  p.setInternal("spread_vs",spread_vs_,getUnits("spread_vs"));
  p.setInternal("spread_alt",spread_alt_,getUnits("spread_alt"));
}

void AlertThresholds::setParameters(const ParameterData& p) {
  if (p.contains("region")) {
    setRegion(BandsRegion::valueOf(p.getString("region")));
  }
  if (p.contains("alerting_time")) {
    setAlertingTime(p.getValue("alerting_time"));
    units_["alerting_time"] = p.getUnit("alerting_time");
  }
  if (p.contains("early_alerting_time")) {
    setEarlyAlertingTime(p.getValue("early_alerting_time"));
    units_["early_alerting_time"] = p.getUnit("early_alerting_time");
  }
  if (p.contains("spread_trk")) {
    setTrackSpread(p.getValue("spread_trk"));
    units_["spread_trk"] = p.getUnit("spread_trk");
  }
  if (p.contains("spread_gs")) {
    setGroundSpeedSpread(p.getValue("spread_gs"));
    units_["spread_gs"] = p.getUnit("spread_gs");
  }
  if (p.contains("spread_vs")) {
    setVerticalSpeedSpread(p.getValue("spread_vs"));
    units_["spread_vs"] = p.getUnit("spread_vs");
  }
  if (p.contains("spread_alt")) {
    setAltitudeSpread(p.getValue("spread_alt"));
    units_["spread_alt"] = p.getUnit("spread_alt");
  }
}

std::string AlertThresholds::getUnits(const std::string& key) const {
  std::map<std::string,std::string>::const_iterator got = units_.find(key);
  if (got == units_.end()) {
    return "unspecified";
  }
  return got->second;
}

}
